client/my-sites/plans-features-main/hooks/use-plan-upgradeability-check.ts

```diff
@@ -5,20 +5,22 @@ import type { PlanSlug } from '@automattic/calypso-products';
 
 interface Props {
 	planSlugs: PlanSlug[];
+	sitePlanSlug: PlanSlug | null;
 }
 
 type PlanUpgradeability = {
 	[ planSlug in PlanSlug ]: boolean;
 };
 
-const usePlanUpgradeabilityCheck = ( { planSlugs }: Props ): PlanUpgradeability => {
+const usePlanUpgradeabilityCheck = ( { planSlugs, sitePlanSlug }: Props ): PlanUpgradeability => {
 	const selectedSiteId = useSelector( getSelectedSiteId );
 	const planUpgradeability = useSelector( ( state ) => {
 		return planSlugs.reduce( ( acc, planSlug ) => {
 			return {
 				...acc,
 				[ planSlug ]:
-					!! selectedSiteId && isPlanAvailableForPurchase( state, selectedSiteId, planSlug ),
+					! sitePlanSlug ||
```

Comment:
Currently, this hook always returns `false` for the upgradeability check for all plans if in onboarding. This seems odd to me, because all plans are selectable ( "upgradeable" ) in onboarding, although I might be missing something.

We can verify whether or not we're in onboarding a few different ways, but I just check for existence of a `sitePlanSlug` as a placeholder for now.

@chriskmnds wondering if you have thoughts on this. This hook appears to only be used [when constructing the gridPlans data structure]( https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plans-grid/hooks/npm-ready/data-store/use-grid-plans.tsx#L311 ), and we actually ignore results from the hook if we're in signup anyway.

https://github.com/Automattic/wp-calypso/blob/8cc3bf3d93671fd9699168bd73099d47c9b00b11/client/my-sites/plans-grid/hooks/npm-ready/data-store/use-grid-plans.tsx#L338

Reply:
> I think semantically if the siteSlug is empty, this hook should return null not true or false. Because if a site slug is not provided the question of whether this site can be upgraded is a non starter. Because the question of whether a site can be upgraded or not always is tied to a site. And on onboarding, a site currently does not exist :D

Dehan and I chatted about this point more. To elaborate, it sounds like `usePlanUpgradeabilityCheck` always expects some sort of meaningful currently selected `siteSlug` or `siteID`, and, in doing so, implicitly conveys that it _will not_ process upgradeability checks in onboarding ( because there is no meaningful selected `siteID` ). Because of this, we'd expect this hook to return `null` instead.

I'm in agreement here, and am happy to add an inline comment and make the change 🙂

Reply:
Really interesting discussion y'all 🙂

> Technically the site Id should be mandatory. And should be made mandatory in types, And if a siteId is not provided we should actually throw an error.

I can understand what you're saying, especially with this precondition "For a given site (with or without a plan) is a given plan purchasable"

> Not entirely sure this is less complex. And just passing the empty site Id to get a true back seems a bit confusing but YMMV. Will resort to your best judgement 

In practice though, for the two use cases ( including this PR ) of `usePlanUpgradeabilityCheck`, we only really care about `true` or `false` values. Returning `true` when there is an undefined `siteId` does, in fact, make existing usage of the hook less complex. Is the plan upgradeable or is it not upgradeable _including_ the consideration of whether or not we're in onboarding?

> return true when plan and undefined siteId are passed

Giving it more I thought, since the two concrete use cases would expect this, I _do_ think it's a reasonable decision to redefine the implicit contract of the hook to account for whether or not we're in onboarding. The only downside I can see is future confusion with unexpected behavior, but this can be mitigated by unit tests / inline documentation.

Reply:
I'd be totally fine with a hook returning true for all plans in Signup when no siteId is passed. Not sure why someone wouldn't be, it conveys "yes, these plans are purchasable when there is no site selected".